docs: update `docs/usage/routing/*` (#3301)
litestar-org/litestar: docs/usage/routing/handlers.rst

==============

Route handlers are the core of Litestar. They are constructed by decorating a function or class method with one of the
handler :term:`decorators <decorator>` exported from Litestar.

For example:

.. code-block:: python
    :caption: Defining a route handler by decorating a function with the :class:`@get() <.handlers.get>` :term:`decorator`

    from litestar import get

    @get("/")
    def greet() -> str:
       return "hello world"

In the above example, the :term:`decorator` includes all the information required to define the endpoint operation for
the combination of the path ``"/"`` and the HTTP verb ``GET``. In this case it will be a HTTP response with a
``Content-Type`` header of ``text/plain``.

What the :term:`decorator` does is wrap the function or method within a class instance that inherits from

> :class:`~.handlers.base.BaseRouteHandler`. These classes are optimized descriptor classes that record all the data

Author: If it helps as an end user reading this, i was lost.

> :class:`~.handlers.base.BaseRouteHandler`. These classes are optimized descriptor classes that record all the data

Reviewer: I think we just remove that terminology - its not doing beginners any favors by being there.

* ``socket``: injects the :class:`WebSocket <.connection.WebSocket>` instance. Available only for `websocket route handlers`_
* ``state`` : injects a copy of the application :class:`State <.datastructures.state.State>`.
* ``body`` : the raw request body. Available only for `HTTP route handlers`_

Note that if your parameters collide with any of the reserved :term:`keyword arguments <argument>` above, you can
:ref:`provide an alternative name <usage/routing/parameters:Alternative names and constraints>`.

For example:

.. code-block:: python
    :caption: Providing an alternative name for a reserved keyword argument

    from typing import Any, Dict
    from litestar import Request, get
    from litestar.datastructures import Headers, State

    @get(path="/")
    async def my_request_handler(
       state: State,
       request: Request,
       headers: Dict[str, str],
       query: Dict[str, Any],
       cookies: Dict[str, Any],
    ) -> None: ...

.. tip:: You can define a custom typing for your application state and then use it as a type instead of just using the
    :class:`~.datastructures.state.State` class from Litestar

Type annotations
----------------

> Litestar enforces strict type annotations. Functions decorated by a route handler **must** have all their

Reviewer: I wonder if there's a link we can use for the term "annotations" like we've done for other things like decorators etc. Maybe https://peps.python.org/pep-0484/#type-definition-syntax?

Author: Im not sure here, i believe i tried but the base PR this comes from is a little old so i may be misremembering.

Author: oh not i didnt try at all or hard enough (or both)
Is this what you mean?
https://docs.python.org/3/glossary.html#term-annotation

Reviewer: Yeh that'll do!

> :term:`kwargs <argument>` and return value type annotated.

Reviewer: ```suggestion
:term:`arguments <argument>` and return value type annotated.
```
I think this is more correct.

If a type annotation is missing, an :exc:`~.exceptions.ImproperlyConfiguredException` will be raised during the
application boot-up process.

There are several reasons for why this limitation is enforced:

#. To ensure best practices
#. To ensure consistent OpenAPI schema generation

> #. To allow Litestar to compute during the application bootstrap all the :term:`kwargs <argument>` required by a
>    function

Reviewer: ```suggestion
#. To allow Litestar to compute the :term:`arguments <argument>` required by a function during application bootstrap
```

HTTP route handlers
-------------------

The most commonly used route handlers are those that handle HTTP requests and responses.
These route handlers all inherit from the :class:`~.handlers.HTTPRouteHandler` class, which is aliased as the
:term:`decorator` called :func:`~.handlers.route`:

.. code-block:: python
    :caption: Defining a route handler by decorating a function with the :class:`@route() <.handlers.route>`
      :term:`decorator`

    from litestar import HttpMethod, route

    @route(path="/some-path", http_method=[HttpMethod.GET, HttpMethod.POST])
    async def my_endpoint() -> None: ...

> As mentioned above, :func:`@route() <.handlers.route>` does is merely an alias for ``HTTPRouteHandler``,

Reviewer: ```suggestion
As mentioned above, :func:`@route() <.handlers.route>` is merely an alias for ``HTTPRouteHandler``,
```
